File: week-2/problem-set/prob9.py
"""Prob9 example.

This file belongs to the week 2 data-structure and graph practice solutions. It
shows two ways to sort a queue while only using queue-like helper structures.
"""


def order_queue_two(original_queue):
    """Sort a queue using two helper queues.

    Args:
        original_queue (list): The original_queue input used by this function.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    working_queue = list(original_queue)
    buffer_queue = []
    sorted_queue = []

    while working_queue:
        current_min = working_queue.pop(0)

        while working_queue:
            value = working_queue.pop(0)
            if value < current_min:
                buffer_queue.append(current_min)
                current_min = value
            else:
                buffer_queue.append(value)

        sorted_queue.append(current_min)
        working_queue = buffer_queue
        buffer_queue = []

    original_queue.clear()
    original_queue.extend(sorted_queue)


def order_queue_one(original_queue):
    """Sort a queue using one helper queue.

    Args:
        original_queue (list): The original_queue input used by this function.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    working_queue = list(original_queue)
    helper_queue = []

    while working_queue:
        current_value = working_queue.pop(0)

        while helper_queue and helper_queue[-1] > current_value:
            working_queue.append(helper_queue.pop())

        helper_queue.append(current_value)

    original_queue.clear()
    original_queue.extend(helper_queue)


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    first_queue = [4, 1, 3, 2, 5]
    order_queue_two(first_queue)
    print(first_queue)

    second_queue = [4, 1, 3, 2, 5]
    order_queue_one(second_queue)
    print(second_queue)


if __name__ == "__main__":
    main()

File: week-2/problem-set/prob9.cpp
/**
 * @file prob9.cpp
 * @brief C++ version of the queue-ordering examples.
 */

#include <deque>
#include <iostream>

using namespace std;

/**
 * @brief Sort a queue using two helper queues.
 * @param original_queue deque<int>& Queue to sort in ascending order.
 * @return void This function updates original_queue and does not return a value.
 */
void order_queue_two(deque<int>& original_queue) {
    deque<int> working_queue = original_queue;
    deque<int> buffer_queue;
    deque<int> sorted_queue;

    while (!working_queue.empty()) {
        int current_min = working_queue.front();
        working_queue.pop_front();

        while (!working_queue.empty()) {
            int value = working_queue.front();
            working_queue.pop_front();
            if (value < current_min) {
                buffer_queue.push_back(current_min);
                current_min = value;
            } else {
                buffer_queue.push_back(value);
            }
        }

        sorted_queue.push_back(current_min);
        working_queue = buffer_queue;
        buffer_queue.clear();
    }

    original_queue = sorted_queue;
}

/**
 * @brief Sort a queue using one helper queue.
 * @param original_queue deque<int>& Queue to sort in ascending order.
 * @return void This function updates original_queue and does not return a value.
 */
void order_queue_one(deque<int>& original_queue) {
    deque<int> working_queue = original_queue;
    deque<int> helper_queue;

    while (!working_queue.empty()) {
        int current_value = working_queue.front();
        working_queue.pop_front();

        while (!helper_queue.empty() && helper_queue.back() > current_value) {
            working_queue.push_back(helper_queue.back());
            helper_queue.pop_back();
        }

        helper_queue.push_back(current_value);
    }

    original_queue = helper_queue;
}

/**
 * @brief Print a queue in bracket form.
 * @param values const deque<int>& Queue to print. Each element has type int.
 * @return void This helper only prints data and does not return a value.
 */
void print_queue(const deque<int>& values) {
    cout << '[';
    for (size_t index = 0; index < values.size(); ++index) {
        cout << values[index];
        if (index + 1 != values.size()) {
            cout << ", ";
        }
    }
    cout << ']';
}

/**
 * @brief Run the queue-ordering demonstrations.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    deque<int> first_queue = {4, 1, 3, 2, 5};
    order_queue_two(first_queue);
    print_queue(first_queue);
    cout << '\n';

    deque<int> second_queue = {4, 1, 3, 2, 5};
    order_queue_one(second_queue);
    print_queue(second_queue);
    cout << '\n';
    return 0;
}
